open-api/src/test/java/org/apache/iceberg/rest/RESTCompatibilityKitCatalogTests.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.rest;
+
+import org.apache.iceberg.catalog.CatalogTests;
+import org.apache.iceberg.util.PropertyUtil;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RESTCompatibilityKitCatalogTests extends CatalogTests<RESTCatalog> {
+  private static final Logger LOG = LoggerFactory.getLogger(RESTCompatibilityKitCatalogTests.class);
+
+  private static RESTCatalog restCatalog;
+
+  @BeforeAll
+  static void beforeClass() throws Exception {
+    if (RCKUtils.isLocalServer()) {
+      RCKUtils.startLocalServer();
```

Comment:
It would be much nicer to use a JUnit5 Extension for managing the server lifecycle. Also, the extension could support pluggable code for non-default servers. This way the test code can be written without any dependency on the server side, while server startup becomes a runtime concern of the test extension.

Reply:
That would be great, but I think this could be split out as a follow on PR.

Reply:
Ok, I was able to move the local server to an extension, but I'm not sure it's really pluggable from a test perspective, but I also don't think that's necessary.  If you take a look at the README, you can disable the local server and configure any external server.  The local server just validates all of the catalog tests against the adaptor based implementation which ensure at least the reference implementation complies.

Reply:
Thx! Annotating leaf test classes should work fine. A test suite targeting another server would be able to define its own leaf test class (extending the base TCK class) and have it annotated with a different extension.

Reply:
I guess it would be nice to separate actual TCK test code, from helper classes for the "local" server in terms of maven artifacts.

Reply:
I tried a few things to separate the tests and local server, but there's always a dependency of some sort (whether it's the JUnit extension or the server classes) on the testFixtures.

The testFixtures are produced as an artifact, so those who want to build client tests against the server can do that (this is what the other projects like pyiceberg/rust do).  For running the tests without the test fixtures, I don't have a great solution now, but I'm open to figuring out a way to separate those.

Reply:
With extensions, I'd usually have it inject (as a JUnit5 `ParameterResolver`) some java object (e.g. `URI`) or config into test classes for them to use in creating the REST Catalog (client)... Pseudo-code for example:
```
static void beforeClass(@RESTServerURI URI uri, @RCKConfig Map<String, String> config) throws Exception {
    restCatalog = RCKUtils.initCatalogClient(uri, config);
}
```
`@RESTServerURI` would be defined in the TCK. Non-default test extension will be responsible for injecting it for whatever custom server impl. The URI can be part of the config, of course.

The code that uses `@RESTServerURI` and `@RCKConfig` does not have to depend on test extensions.

For example, RCK logic is in class `RESTCompatibilitySuite` (abstract, for reuse) - no references to system props, env, or `RESTServerCatalogAdapter`. Then:
```
@ExtendWith(RESTServerExtension.class)
class RESTCompatibilityReferenceTest extends RESTCompatibilitySuite {
 // this class is empty and NOT REUSED outside of the Iceberg repo
}
```
Loading config from system props and env is handled by `RESTServerExtension`, which also injects config into `RESTCompatibilitySuite`.

WDYT?

Reply:
I like this in concept, but in practice, I wasn't able to find an approach that works.  One of the issues is that the Suite does not work with the `@ExtendWith` annotation, so extending the suite class does not allow for extension like in your example. This combined with the multiple test classes we need to extend (`CatalogTests<RESTCatalog>` and `ViewCatalogTests<RESTCatalog>`) means that we don't have an easy way to make this pluggable/extendable.

I think we probably want to have a set of standardized tests that we package here and allow for reasonable feature flag tuning, but if you want to customize tests beyond what's defined, just create a separate suite of tests.